Update llvm-6.x to llvm-13.x: use the Orc Jit API.

CiaranWelsh opened the pull request:
- This is the LLVM update work. 
- Future builds will need to use [llvm-13.x](https://github.com/sys-bio/llvm-13.x) rather than llvm-6.x. 
- Adds the `Jit` hierarchy - MCJit (old API) and LLJit (llvm's out of the box front end for Orc Jit API). 
- `Jit` is injected as a dependency into `LLVMModelGenerator` and handles the compilation of the roadrunner model. 
    - It is therefore extensible, and future work should look into LLJit's "Lazy compilation" feature as another `Jit`
-  We have a new option in `Config` for switching between backends. 
    - `Config::setValue(Config::LLVM_BACKEND, Config::LLJit); // or MCJit`
    - And for choosing number of threads, which is a feature of LLJit 
        - `Config::setValue(Config::LLJIT_NUM_THREADS, 1);`
- Some things can likely be improved but on balance it is better it merge this sooner rather than later. i.e. 
    - It feels like the `ModelGeneratorContext` could be entirely replaced by the new `Jit`. 
    - Interface segregation principle is broken in the `Jit`, because it was hard to reconcile MCJit and LLJit which are two different API's to achieve the same goals.  It works however, so there's that... 
- Beginnings of a performance test suite
- Updated version from v2.1.4 to v2.2.0, because of the significant backend changes.

Commit:
Add some comments.

Commit:
Turned some raw pointers into unique pointer as they are passed as arguments to some newer llvm api's.

Commit:
Add orc jit tollvm libraries

Commit:
make engine argument nullptr. Its null anyway, this is clearer.

Commit:
remove code duplication

Commit:
Remove the executionEngine member variable of CodeGenBase. It is never actually used and it was replaced in modern llvm with an ExecutionSession - therefore this variable complicates building an abstract interface that supports MCJit and OrcJit.

Commit:
Begining of a Jit hierachy for encapsulating MCJit and OrcJit operations. MCJit works in isolation of roadrunner for compiling a fibonacci function. Need to commit to switch branches and print an example of complete IR code to compare with.

Commit:
Add tests for math functions that are pulled out of the C library by MCJIt

Commit:
Add more tests for functions obtained from "elsewhere". Add function pointers to these functions in base class Jit.

Commit:
successfully encapsulated the old style llvm-6 jit configuration in a MCJit class. Tests pass. This may need to change in order to support dependency injected Jit engine into RoadRunner object. Certainly a good start

Commit:
refactor JitTests for code reuse

Commit:
run brown2003_multoprocessing_pool script to get a feel for whether disabling the function pass manager per added function increases build speed. The idea is that we already use the PassManager at the module level, so maybe doing it again at the function level is redundant. Initial tests indicate a x5 speedup

Commit:
Created an interface in Jit for the process of adding symbols and mapping them to functions. This used to be called addGlobalMappings in older roadrunner versions but this has changed to mapFunctionsToJitSymbols to be a little more presise about what the function is doing and so that it applies to other Jit types. Tests for all 40 functions whether acquired from the standard C library or from roadrunner sbmlsupport namespace for both Jit types.

Commit:
Add a key in Config for switching between llvm backends

Commit:
Add ability to create BasicDictionary with initializer list.

Commit:
Added support for switching llvm backend compilers from both Config and LoadSBMLOptions. Tests in LoadSBMLOptionsTests.cpp.

Commit:
Got the switch between Jit versions working. rrLLJit is still has problems.

Commit:
Tried to move some things around but broke something. Commit to new branch to investigate

Commit:
uncomment out section that was needed for MCJit

Commit:
Move mStreamOut to Jit superclass. This is a buffer used to store llvm IR in and to make an object file.

Commit:
fixed segfault that resulted from not transferring new objects properly between ModelgeneratorContext and ModelResources

Commit:
Merge branch 'orc-jit-seg-fault' into orc-jit

# Conflicts:
#	source/llvm/Jit.h
#	source/llvm/LLVMModelGenerator.cpp

Commit:
Added some unit tests for distrib. These are stochastic in nature so they are quite weak tests.. but its still useful to check whether the functions are loaded correctly, etc...

Commit:
Refactor

Commit:
Commit work before experimenting with changing constness

Commit:
Change Internal to External linkage in the call to llvm::Functino::Create. Otherwise orcjit cannot locate memory address by name.

Commit:
The call to addIRModule steals the reference to module, causing a segfault. Jiggle around some calls so that the call addIRModule happens last.

Commit:
LLJit works but MCJit segfaults. Commit before experimental changes to dig into the problem.

Commit:
Fixed segfault in MCJit.

MCJit converts post optimized stream to an object file, whereas LLJit does not. We will probably need to reconsile these two ways of doing things when I find bugs.

Commit:
Add checks for empty stringified roadrunner object file in loadstate.

Commit:
Bugfix - move check that moduleStr in loadState is not empty to after the moduleStr variable has been actually loaded.

Commit:
clean up, all tests pass with MCJit, now need to figure out whether they work with LLJit

Commit:
factored out some duplicated code - the bits that are shared between LLVMModelGenerator::createModel and LLVMModelGeneratoe::regenerateModel. The LLJit was developed using createModel as test code but duplicated code meant that these changes were not realised in regenerate model... and segfault.

Commit:
move createCLibraryFunctions from MCjit to Jit since they are also needed in LLJit. This fixes problem with simulating venkatraman model which uses th epower function£

Commit:
Added function to dump symbol table from Jits

Commit:
Added function to dump symbol table from LLJit only.

Commit:
Implemented a caching system for LLJit. This is a subclassed LLVM structure rather than our home rolled version.

Commit:
Refactor

Commit:
Added test that grabs an object file from cache and loads that rather than needing to compile.

Commit:
Bugfix with setting the name of the module - use moduleNonOwning not module which was stolen by ThreadSafeModule

CiaranWelsh committed:
Added linear chain performance test in python.

CiaranWelsh committed:
Remove code duplication in loadState by reusing the code in MCJit for loading function pointers for ModelResources

CiaranWelsh committed:
Remove superflous duplicated code.

CiaranWelsh committed:
With LLJit, get the module string directly from the object cache, after it has been created and cached.

CiaranWelsh committed:
Implement a JitFactory that uses either global Config options or options from LoadSBMLOptions::modelGeneratorOpt

CiaranWelsh committed:
Begin testing some llvm generated code.

CiaranWelsh committed:
python tests and load/save state tests pass again after breaking them for MCJit. The problem is that I tried to take a short cut when creatin gthe binary from the input stream.

CiaranWelsh committed:
All tests pass. Windows, locally.

CiaranWelsh committed:
Refactor rc to modelResources. Add comments to Jit.

CiaranWelsh committed:
Clean up saving the module string for a potential saveState later. Instead of remembering to save the module string in multiple locations through model generation, we have a method in the Jit interface that gives us the module binary string. This is called in saveState, directly when needed.

CiaranWelsh committed:
Delete old OrcJit class

CiaranWelsh committed:
Clean up

CiaranWelsh committed:
Added performance test suite

CiaranWelsh committed:
Add to performance test suite -- N reactions with inceasing connectivity

CiaranWelsh committed:
fixed memory bug in performance test where antimomny holds on to too many models, consumed memory and then presumably slows future computation

CiaranWelsh committed:
Run performance test suite 600

CiaranWelsh committed:
Refactor LLVM_COMPILER to LLVM_BACKEND for choosing between MCJit and LLJit

CiaranWelsh committed:
LLVMCompiler => LLVMBackend

CiaranWelsh committed:
Added option to control LLJit optimization level

CiaranWelsh committed:
Added option for LLJit_NUM_THREADS in global Config. Note, we have not added this option to the LoadSBMLOptions.modelGeneratorOpt, even though it belongs there because this value is an integer and modelGeneratorOpt uses bitfields and bitmasking to toggle features on and off. In short, to change the number of threads in use for the LLJIT backend, you must use the global Config system, not LoadSBMLOptions.

CiaranWelsh committed:
Refactor NReactions to NSpecies. Makes more sense.

CiaranWelsh committed:
Add biomodels test

CiaranWelsh committed:
Add biomodels test

CiaranWelsh committed:
Save binary stream of compiled llvmIR module in saveState.

CiaranWelsh committed:
Numerious problems were fixed in save state and the original save state test suite. The main issue was where the binary stream containing the object file was not saved with saveState. This problem doesn't matter until you try to save the state of a copy of a roadrunner model. Copying a roadrunner model also uses save state, so by not transferring the binary stream containing the compiled model caused segfault when trying to access the stream to convert the binary stream to string (in MCJit, LLJit was fine).

This exposed some problems with the save state test suite. The main issue is that the generic string "save-state-test.rr" was used for a file path. Where exactly c++ decided to save this I don't know, but I suspect in home or somewhere similar. This means that the first time the save state tests are run, this file is generated and then reused for every test! Tests must be independent and therefore this is a bug. Fixing this problem revealed several tests that loadState before having first saved the state. These have been disabled.

CiaranWelsh committed:
add biomodels.zip to gitignore. Remove it from git ==> large file

CiaranWelsh committed:
only save compiledModuleBinaryStream when it is populated, aka in MCJit mode

CiaranWelsh committed:
Run the performance test suite on random mass action and Ham... kinetics networks. Store the data.

CiaranWelsh committed:
add some models for performance testing

CiaranWelsh committed:
add barplots for simulation time data and remove inline from some methods calls since msvc generators don't like them (NMake seems to be fine however)

CiaranWelsh committed:
Problem with long test name using msvc generator. Sad face =[ Fix python test getSBML not getSBMLFIle

CiaranWelsh committed:
Update install tree so that it can be used from c++

CiaranWelsh committed:
BuildAModel.cpp added for use with intels VTune profiler.

CiaranWelsh committed:
Stack allocate ModelGeneratorContext.cpp rather than heap.

CiaranWelsh committed:
Updated dockerfile to build llvm-13.x. Renamed latest tag to llvm-6.x so you can download the old roadrunner docker image with `docker pull ciaranwelsh/roadrunner-manylinux2014-base:llvm-6.x` and the new one with `docker pull ciaranwelsh/roadrunner-manylinux2014-base:llvm-13.x`

First attempt at updating the azure pipelines file as well. Remove debug builds and only build the full workflow with Python. This is better because of limited resources and we are currently running the C++ tests twice.

CiaranWelsh committed:
update swig link in dockerfile

CiaranWelsh committed:
remove typo in azure

CiaranWelsh committed:
turn off llvm caching on azure - we need to redownload it

CiaranWelsh committed:
turn off llvm caching on azure - we need to redownload it

CiaranWelsh committed:
turn off llvm caching on azure - we need to redownload it

CiaranWelsh committed:
copy llvm13 into the llvm folder for linux. Add execute permissions for lvm-config on mac and fix swig on windows

CiaranWelsh committed:
comment out swig cache on windows

CiaranWelsh committed:
remove quotes linux build

CiaranWelsh committed:
remember to checkout orc-jit branch in azure

CiaranWelsh committed:
Add default CMAKE_INSTALL_PREFIX. This variable is required in roadrunner as it is used throughout the build system. If not provided by the user, we provide a default.

CiaranWelsh committed:
comment out llvm function that is only available in debug builds

CiaranWelsh committed:
Add python file for getting llvm from git repo.

CiaranWelsh committed:
use python 2 syntax

CiaranWelsh committed:
conda on azure non python builds

CiaranWelsh committed:
install requests

CiaranWelsh committed:
wrong directory for llvm-config permissions

CiaranWelsh committed:
use llvm python script for windows and ubuntu as well

CiaranWelsh committed:
windows azure add miniconda to non python build

CiaranWelsh committed:
add PythonName variable to windows non-python build

CiaranWelsh committed:
add PythonVersion variable to windows non-python build

CiaranWelsh committed:
install conda before using py39 windows non-python builds

CiaranWelsh committed:
remove BuildAModel tmp executable

CiaranWelsh committed:
fixed the libroadrunner-deps package which has been rebuilt and new artifacts available for the roadrunner azure pipeline to download

CiaranWelsh committed:
updated dockerfile manylinux2014

CiaranWelsh committed:
checkout orc-jit branch on non python builds

CiaranWelsh committed:
download dbg llvm not release for windows

CiaranWelsh committed:
missing cpp file in one of the test binaries

CiaranWelsh committed:
remove llvm cache for windows build

CiaranWelsh committed:
install requests miniconda on linux non python build

Commit:
LLVM's MCJit on macos adds _ char to externally defined functions. Put preprocessor macro in place to detect __APPLE__ and add the leading underscore

Commit:
change manual adding '_' on macos to formal name mangling method from LLVM

Commit:
MCJit works on macos. Problem with name mangling. Still working on LLjit problem with similar issue.

Commit:
changed the way LLJit maps functions to Jit symbols. This works on mac, untested on windows/linux

Commit:
change the azure pipelines download libroadrunner-deps task such that the libomexmeta's master branch is the only one used

CiaranWelsh committed:
uncomment out sbml test suite tests

CiaranWelsh committed:
Merge branch 'develop' into orc-jit-merge-develop

Fixed conflicts between orc-jit and develop branch.

# Conflicts:
#	azure-pipelines.yml
#	source/ExecutableModelFactory.cpp
#	source/SteadyStateSolver.h
#	source/llvm/LLVMModelGenerator.cpp
#	source/llvm/LLVMModelGenerator.h
#	source/llvm/ModelGeneratorContext.cpp
#	source/llvm/ModelGeneratorContext.h
#	source/rrRoadRunner.cpp
#	test/cxx_api_tests/RoadRunnerAPITests.cpp
#	test/serialization/state_saving.cpp
#	wrappers/C/rrc_api.cpp

CiaranWelsh committed:
fixed tests and a few bugs after merging orc-jit and develop

CiaranWelsh committed:
change master to release

CiaranWelsh committed:
quote release in azure pipelines

CiaranWelsh committed:
release ==> refs/heads/release

CiaranWelsh committed:
add llvm caching back in on CI. Remove 'ModelGeneratorContext::ModelGeneratorContext' from decl

CiaranWelsh committed:
Linux ==> Ubuntu on CI. Swap values of PythonName and PythonVersion on CI

CiaranWelsh committed:
comment out temporary test that only works locally

CiaranWelsh committed:
use bash/powershell to selectively run the tests only in Release

CiaranWelsh committed:
update llvm link for windows - looked like a Debug vs Release mixup

CiaranWelsh committed:
update version

CiaranWelsh committed:
Quote the if statement

CiaranWelsh committed:
remove orc-jit branch from azure, rely instead on the PR@

luciansmith committed:
Attempt to fix max fail.

Fail on read file might be because it relies on vagaries of how tests are run, loading a save file written elsewhere.  This version makes everything self-contained, but still checks 'load on new instance'.

luciansmith committed:
Trim azure.

* Don't build Ubuntu Python; we already have manylinux.
* Don't build for Py3.6
* Don't run tests twice on manylinux/Python; just run the python tests (we run the normal tests already on Ubuntu).

These changes had been in develop, but were difficult to merge; this re-implements the changes.

CiaranWelsh committed:
fix typo in running tests only in release mode on windows

CiaranWelsh committed:
comment out python_tests_RunStochasticTestSuite as it takes too long on many linux -- does the build succeed without this test? If so, we need to find a way to shorten python_tests_RunStochasticTestSuite

CiaranWelsh committed:
comment out llvm caching on windows cpp build. Force to download the updated llvm

CiaranWelsh committed:
comment out llvm caching on windows cpp build. Force to download the updated llvm

CiaranWelsh committed:
debug build on windows fails because of 10GB size limit on azure. Therefore we stop building the tests on azure/debug/win32

CiaranWelsh committed:
no cache llvm

CiaranWelsh committed:
add BuildTests variable to windows cpp build. Do not build tests for debug builds

luciansmith commented on test/CMakeLists.txt:
```diff
@@ -316,11 +316,12 @@ if (BUILD_PYTHON AND BUILD_TESTS)
     #add_pytest_to_ctest(python_tests_StructuralAnalysisTests "${RR_PYTHON_TESTING_BUILD_PREFIX}/test_structuralAnalysis.py")
 
+    # add_pytest_to_ctest(python_tests_RunStochasticTestSuite "${RR_PYTHON_TESTING_BUILD_PREFIX}/test_runStochasticTestSuite.py")
+
```
This isn't a good sign--these tests easily ran in under an hour before the switch to LLVM 13.  However, there was a single test that was inexplicably slow in this suite that I commented out--maybe with a slightly different execution path, there's a different test or two that is inexplicably slow now?  I'll investigate.

luciansmith replied:
https://dev.azure.com/TheRoadrunnerProject/255e6ce3-64bc-4a95-a8c8-2a25dc15a90e/_apis/build/builds/1158/logs/183
took 18 minutes.

https://dev.azure.com/TheRoadrunnerProject/255e6ce3-64bc-4a95-a8c8-2a25dc15a90e/_apis/build/builds/1244/logs/242
took >40 and timed out.

CiaranWelsh replied:
I've double checked the [RandomTests](https://github.com/sys-bio/roadrunner/blob/6325f0c9ac09740787ea17625918bbe0455b0eb7/test/llvm_tests/RandomTests.cpp) and they run in an expected amount of time. I'm not sure why some tests hang, I wonder if its a similar problem that we had with the pickle?

CiaranWelsh replied:
I ran python_test_runStochasticTests.py just locally (windows) and got this:

```
  File "C:\Miniconda3\envs\py39\lib\unittest\case.py", line 592, in run
    self._callTestMethod(testMethod)
  File "C:\Miniconda3\envs\py39\lib\unittest\case.py", line 550, in _callTestMethod
    method()
  File "D:\roadrunner\roadrunner\test\python\test_runStochasticTestSuite.py", line 322, in test_stoch_00014
    self.runOneTest('00014')
  File "D:\roadrunner\roadrunner\test\python\test_runStochasticTestSuite.py", line 240, in runOneTest
    (nmean_wrong, nsd_wrong, nlnmean_wrong, nlnsd_wrong) = self.simulateFile(file, tnum, expected_results)
  File "D:\roadrunner\roadrunner\test\python\test_runStochasticTestSuite.py", line 189, in simulateFile
    sim = rr.simulate(self.start, self.duration, self.steps+1)
  File "D:\roadrunner\roadrunner\cmake-build-release\lib\site-packages\roadrunner\roadrunner.py", line 4603, in simulate
    result = self._simulate(o)
  File "D:\roadrunner\roadrunner\cmake-build-release\lib\site-packages\roadrunner\roadrunner.py", line 4269, in _simulate
    return _roadrunner.RoadRunner__simulate(self, opt)
KeyboardInterrupt
sys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name='../../test/sbml-test-suite/stochastic/results.tsv' mode='w' encoding='cp1252'>

```

CiaranWelsh replied:
Okay, possible fix just pushed -- using `for line in open(fname)` ensures that there is no way to close the file again.

luciansmith replied:
I updated your fix with calls to f.close(), and I also commented out the newly-slow test (23).  Fingers crossed...

I've also tried running this using my linux VM, and it's EXTREMELY slow there, both with librr 2.1.3 and with the latest (so at least llvm didn't cause problems there).  It looks like it runs in approximately the same amount of time on Windows in Azure, too, which is nice, though it's still confusing about manylinux, since it worked fine there with llvm6.

luciansmith replied:
(Also, the 'open' issue was present before when it ran fine, so I wouldn't think it would be the cause of the problem.)

luciansmith replied:
I'm filing this as https://github.com/sys-bio/roadrunner/issues/929 and closing this issue.

CiaranWelsh commented:
I tried to run these tests on many linux
```
docker pull ciaranwelsh/roadrunner-manylinux2014-base:llvm13.x
docker image ls
docker run -ti <insert hash of image>
# image runs in interactive mode
bash build-roadrunner.sh # custom bash script
```
Then I tried to run the stochastic test suite from the installed Release and got 
```
Traceback (most recent call last):
  File "/Roadrunner-release/site-packages/roadrunner/tests/test_runStochasticTestSuite.py", line 27, in setUpClass
    cls.fail("Unable to find stochastic test suite directory.")
  File "/Miniconda3/envs/py39/lib/python3.9/unittest/case.py", line 668, in fail
    raise self.failureException(msg)
AttributeError: 'str' object has no attribute 'failureException'
```
I think we need to install the missing files in a cmake command?

CiaranWelsh committed:
remove ifdef that was preventing RandomTests from running with LLJit

CiaranWelsh committed:
use pythons context manager for handling open commands

CiaranWelsh committed:
uncomment out python_tests_runStochasticTestSuite

luciansmith committed:
Close the files; print more information.

luciansmith commented:
> Traceback (most recent call last):
>   File "/Roadrunner-release/site-packages/roadrunner/tests/test_runStochasticTestSuite.py", line 27, in setUpClass
>     cls.fail("Unable to find stochastic test suite directory.")
>   File "/Miniconda3/envs/py39/lib/python3.9/unittest/case.py", line 668, in fail
>     raise self.failureException(msg)
> AttributeError: 'str' object has no attribute 'failureException'

I haven't been able to figure out how to call 'fail' properly in the 'setUpClass' routine; it's supposed to just fail with the message in question but instead behaves as above.  It still fails, though, so I left it for now.

The failure means the test can't find the test suite directory.  It works from the source/test/python directory, and it works from the build directory in the tests; I haven't tried to install the tests with everything else yet so it would work from the install directory.  As you say, this could be done with cmake.

CiaranWelsh commented:
> I haven't been able to figure out how to call 'fail' properly in the 'setUpClass' routine;

I think fail is used internally in unittest to implement the assert statements. Normally people use the assert statements. [see here for cheat sheet](https://kapeli.com/cheat_sheets/Python_unittest_Assertions.docset/Contents/Resources/Documents/index).

luciansmith committed:
Do some testing in azure to give more feedback about manylinux timeout.

Modify stochastic test suite to output more information including per-test time.
Remove results file; shouldn't be in source.

luciansmith committed:
Fix stoch test call by cd'ing to site-packages build dir.

luciansmith committed:
Fix command line.

luciansmith committed:
Try local install before running test stoch script.

luciansmith committed:
Install rr when building wheel.

luciansmith committed:
Try ONCE AGAIN to do a local install so I can run this stupid test.

luciansmith committed:
Try a --user install?

luciansmith committed:
It worked!  Add printed information at the very beginning, since nothing happened at all!

luciansmith committed:
Only do half as many tests on linux.

luciansmith committed:
Fix vector return.

CiaranWelsh committed:
Add thread-pool. Begin writing roadrunner container for parallel loading etc.

Commit:
Move lock_guard to include statically defined flag in registrations. These are singleton classes and should be constructed by one thread at a time

luciansmith commented:
So, several things from researching stuff yesterday:
* You can't assert or fail or anything in the setUpClass function; people think it's wrong. and you can't do it.  So, I'm leaving that bit alone.
* I ran the test on my own ubuntu VM, where it also was very slow, and discovered that the new version is generally faster than rr2.1.3:  running it in 2.1.3 took 165 minutes to run the simulations, and 2.2.0 took 155 minutes.  So that's not the problem!
* I was *finally* able to figure out a way to get the stochastic test to actually output stuff to the log, and discovered that the problem is indeed just that it's much, much slower on that platform than anything else.  I think the problem is that the newer manylinux VM is slower than the previous llvm-6-era manylinux VM.  (And also slower than windows/mac, which are also ~15 minutes each.)

luciansmith committed:
Get the more detailed information about the platform so we can skip tests only for Azure.

luciansmith committed:
Revert pipeline and update stoch test.

Now that I finally have the information I need, I can revert the azure pipeline to do what it used to do before I started messing with it.  Then I can update the stochastic test suite to only run a handful of tests on manylinux-azure, which is super slow for unknown reasons.

luciansmith committed:
Fix no-tests for Windows debug build.

Commit:
progress with RoadRunnerMap -- the issue was that the main thread was destructing while other threads were working

luciansmith commented:
OK!  I have it mostly working!  Still don't know why stochastic tests are so slow on manylinux, but I've short-circuited them so we only run one file for 20 of them instead of two files for all 100.

We seem to be running out of disk space for the windows debug build, despite turning off building tests, which is weird, since it worked before.   (Also, if there's any way to continue running windows debug tests, that'd be great--they've found errors in the past.  Maybe we can ditch the pdb files from the deps or something.)

CiaranWelsh commented:
What a faff, I'll try llvm in RelWithDebInfo mode and see what happens. If that fails, we could try deleting the build tree before the final install. This might actually allow us to run the tests, but from the install tree.

CiaranWelsh committed:
Commit to move to macos and use address/thread sanitizer

CiaranWelsh committed:
LLJit

CiaranWelsh committed:
try relwithdbginfo for windows

luciansmith committed:
Switch to using current branch.

Don't checkout the orc-jit branch every time any more.  (Also, slight typo fix)

luciansmith committed:
Use the debug build.

Can't use relWithDebInfo; still get mismatch.

luciansmith commented:
Yes!  Finally got it.  It turns out if you delete all the files in llvm/bin/ except for llvm-config.exe, you reduce the size of the files by about half, which was just enough to let the original debug zip file squeak in under the wire.  It looks like you made the llvm packages by hand instead of from azure?  At any rate, after mucking about with azure for a while, I finally just downloaded the existing zip files, removed the extra exe's and re-uploaded them, and that worked.

I still haven't *reviewed* the changes you made, but at least now everything compiles!  Yay!

CiaranWelsh commented:
I [tried](https://dev.azure.com/TheRoadrunnerProject/llvm-13.x/_build/results?buildId=1159&view=results) to use azure but llvm-13.x debug mode requires >10GB as well, therefore I fell back on manual mode. 

I tried using RelWithDbgInfo for llvm, which reduced the size by about half but something got currupted along the way. Glad you finally got this working though!

Commit:
Refactored the naming of tests for loading jit symbols. Added some unit style tests for loading function but from the module rather than using the jit::getFunctionAddress api. This is in search of a bug that I've found, except all tests pass so I haven't found the problem.

Commit:
I've noticed a problem with LLJit and a potential way to "smooth" over the MCJit and LLJit interface. The problem is that external functions (like arccot) are needed as llvm::Function* types while existing code only enables loading them from the resulting shared library as a pointer. Moreover, it looks like addGlobalMapping actually doesn't do anything in MCJit.

After this commit, I'm going to fix this problem and make some refactorings to try and separate compiler and code generation.

CiaranWelsh committed:
Merge branch 'orc-jit-rr-container' of https://github.com/sys-bio/roadrunner into orc-jit-rr-container

# Conflicts:
#	test/cxx_api_tests/RoadRunnerMapTests.cpp

CiaranWelsh committed:
streamlined the creation of functions like arccot and distrib functions by pulling their defining information into the Jit class which can be used by subclasses to add to the developing shared library

CiaranWelsh committed:
Finish fixing bug where llvm::Module was not able to properly grab llvm::Functions* when compiling the AST.

CiaranWelsh committed:
Merge branch 'orc-jit-rr-container' into orc-jit

CiaranWelsh committed:
Merge branch 'orc-jit' of https://github.com/sys-bio/roadrunner into orc-jit

luciansmith commented on docs/doxygen-output/html/SBMLSupportFunctions_8h_source.html:
```diff
@@ -93,21 +93,21 @@
 <div class="line"><a name="l00038"></a><span class="lineno">   38</span>&#160; </div>
-<div class="line"><a name="l00044"></a><span class="lineno">   44</span>&#160;<span class="keywordtype">double</span> arccot_negzero(<span class="keywordtype">double</span> a);</div>
+<div class="line"><a name="l00044"></a><span class="lineno">   44</span>&#160;<span class="keywordtype">double</span> rr_arccot_negzero(<span class="keywordtype">double</span> a);</div>
 <div class="line"><a name="l00045"></a><span class="lineno">   45</span>&#160; </div>
```
Just to check here to make sure there's a reason arccot_negzero was renamed to rr_arccot_negzero, but the other functions ('cot', 'sec', 'arccot2') were not.

CiaranWelsh replied:
The problem was that I couldn't get the LLJit to locate symbols until they exactly matched the function name - aka ones in sbmlsupport::* .

CiaranWelsh replied:
We could make these more consistent, but would need to rename both [here](https://github.com/sys-bio/roadrunner/blob/32f589ad4ba1ed74bd9c7c6839a045cbbf04800d/source/llvm/Jit.cpp#L320) and [here](https://github.com/sys-bio/roadrunner/blob/15523ffce3d03971eecb229a2dc65b0540f36d87/source/llvm/SBMLSupportFunctions.cpp#L108).

luciansmith replied:
Ah, sounds like the naming issue is something else entirely, then.  Let's not worry about it.

luciansmith commented on docs/doxygen-output/html/classrrllvm_1_1ModelDataIRBuilder.html:
```diff
@@ -281,10 +281,10 @@
 <tr class="memitem:aa128aa56d7a80229166d0466db3ea013"><td class="memItemLeft" align="right" valign="top"><a id="aa128aa56d7a80229166d0466db3ea013"></a>
-static const char *&#160;</td><td class="memItemRight" valign="bottom"><b>csr_matrix_set_nzName</b> = &quot;rr_csr_matrix_set_nz&quot;</td></tr>
+static const char *&#160;</td><td class="memItemRight" valign="bottom"><b>csr_matrix_set_nzName</b> = &quot;csr_matrix_set_nz&quot;</td></tr>
 <tr class="separator:aa128aa56d7a80229166d0466db3ea013"><td class="memSeparator" colspan="2">&#160;</td></tr>
```
It also seems odd to me that these functions would be renamed to *remove* the 'rr', while the previous functions (arccot_negzero) were renamed to *add* it.

luciansmith commented on source/llvm/CodeGenBase.h:
```diff
@@ -40,7 +43,7 @@ class CodeGenBase
         if(functionPassManager)
         {
-            functionPassManager->run(*func);
+            //functionPassManager->run(*func);
```
If this call indeed turned out to not be needed, let's remove the entire if{} statement.

CiaranWelsh replied:
I'm not yet sure whether this is needed or not. We already have module level optimizations, so do we also need function level optimizations? Also, what impact do they actually make? More passes requires more build time, but for how much gain? I left this commented out as a reminder to do some profiling with and without.

luciansmith commented on source/rrLogger.h:
```diff
@@ -341,14 +341,25 @@ RR_DECLSPEC Poco::Logger &getLogger();
+#define rrLogDebug     rrLog(Logger::LOG_DEBUG)
 
+/**
+ * Temporary logging commands to differentiate mine from the original.
```
Do these need to be deleted now?

luciansmith replied:
Actually, that was a dumb question.  Of course they should be removed; I'll do it.

luciansmith commented on test/PerformanceTests/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+
+
+
+set(test_targets ${test_targets} PARENT_SCOPE)
```
@CiaranWelsh : Do you want the OldData pngs checked in?  If so that's fine; just wanted to be sure you wanted 'em.

CiaranWelsh replied:
Nope, actually I might get rid of most of these python tests as I'm currently working on a better way by wrapping vtune CLI commands and running them with python. It remains to be seen whether we can get this running on Azure, but at least we've automated a tool that specializes in performance tests. I'm thinking a new Azure pipeline to sit alongside our main pipeline might work well, though my credentials problem still exists.

luciansmith commented on test/cxx_api_tests/LoadSBMLOptionsTests.cpp:
```diff
@@ -49,61 +49,202 @@ public :
+//    LoadSBMLOptions opt;
+//    opt.setItem("LLJit_NUM_THREADS", 6);
+//    ASSERT_EQ(6, opt.getItem("LLJit_NUM_THREADS").getAs<int>());
+}
```
This test is entirely commented out.  Do you need it?

CiaranWelsh replied:
No I've now deleted this on my latest (uncommited) branch. The problem was `LoadSBMLOptions` cannot hold anything other than boolean, since it uses bitmasks. Therefore, we fall back on using the global `Config` options instead of allowing the user to pass in the option using `LoadSBMLOptions`. This is actually a bit of a shame because the other build options are all accessible via `LoadSBMLOptions`.

luciansmith commented on test/cxx_api_tests/RoadRunnerAPITests.h:
```diff
@@ -0,0 +1,1734 @@
+//
```
I can't figure out why you renamed this file from .cpp to .h

luciansmith replied:
Got it!  Seems good to me.

luciansmith commented on test/cxx_api_tests/RoadRunnerAPITestsWithLLJit.cpp:
```diff
@@ -0,0 +1,859 @@
+
+
+TEST_F(RoadRunnerAPITestsWithLLJit, DefaultJacobianMode){
+    DefaultJacobianMode();
```
This test and the others in this file look like they are prototypes and not full tests yet.  True?

luciansmith replied:
Nope, figured it out--you renamed the other test .h so that you could subclass these tests with different options.  Got it!

luciansmith commented on test/cxx_api_tests/RoadRunnerMapTests.cpp:
```diff
@@ -0,0 +1,258 @@
+            throw std::invalid_argument("start + N must be less than or equal to 1809: " + std::to_string(end));
+        }
+        std::vector<std::string> vec(N);
+        for (int i = start; i < end; i++) {
```
There are actually a lot of STS tests without the full gamut of l2v1->l3v2 versions.  There are tests for stuff that was only added in l3; there are tests for stuff that was added in l2v2; there are tests for stuff that was removed in l3, etc. etc.  

In general, I set things up to test the earliest and the latest versions of all the tests (keeping in mind that sometimes there's only one test total).  It looks like here you just want the latest?  If so, you at least should remove the warning lines; they're to be expected.

luciansmith replied:
(Actually, I'm removing the warnings myself.  If you indeed only care about grabbing the last one, we're good.)

CiaranWelsh replied:
Okay no problem. Since I needed access to the STS in another test suite I've actually moved this content to a little class called `STS` in `<roadrunner-root>/test/sbml-test-suite`. I've only built enough to support my own tests (so will need expanding in future), but in general I think its a good idea to abstract out access to these models to make it easy to quickly grab them. See my later commit.

luciansmith commented on test/cxx_api_tests/RoadRunnerMapTests.h:
```diff
@@ -0,0 +1,258 @@
+    }
+
+    std::vector<std::string> getFirstNModelsFromSTS(int N, int start = 1) {
+        if (start < 1){
```
Why define this both here in the .h file and then again in the .cpp file?

luciansmith replied:
And indeed, the .h file and the .cpp file are identical.

CiaranWelsh replied:
So, I made my last commit quite hastily because I know you would be reviewing the content and I wanted you to have the latest bug fix. Therefore, these got committed by mistake, in transit to `<roadrunner_root>/test/sbml-test-suite/STS`.

luciansmith commented on test/llvm_tests/FunctionMapping/LLJitMapFunctionsToJitSymbolsTests.cpp:
```diff
@@ -0,0 +1,473 @@
+    std::string funcName = "fabsd";
+    llvm::Function* f = jit->getModuleNonOwning()->getFunction(funcName);
+    std::cout << f << std::endl;
+//    fabsFnTy fnPtr = (fabsFnTy) jit->getModuleNonOwning()->getFunction(funcName);
```
Looks like this test wasn't quite finished.

CiaranWelsh replied:
Sometimes I create dud test for looking into a bug. It's quicker than needing a whole new file with its own main. This did indeed get forgotten about and is now removed.

luciansmith commented on test/llvm_tests/LLVMModelDataSymbolsTests.cpp:
```diff
@@ -0,0 +1,58 @@
+}
+
+TEST_F(LLVMModelDataSymbolTests, r){
+
```
Unfinished test?

CiaranWelsh replied:
An unfinished test suite really. I'm yet to figure out a good way to test some of the internal objects, including this. I'll clean it up a bit but I'd like to keep it as I'm likely to expand more when I get to a point where I need to understand how some of the other objects work. For example, the resolver types might be useful in solving #928

luciansmith replied:
That sounds fine by me!  Maybe make an issue to track the unfinished tests, just so we don't forget in the future.

luciansmith commented on test/llvm_tests/tmp.py:
```diff
@@ -0,0 +1,16 @@
+from roadrunner import RoadRunner, Config
```
The fact that this file is called 'tmp.py' makes me think it might not need to be in the source code?

CiaranWelsh replied:
Yep, I've deleted it on my branch.

luciansmith commented on test/serialization/state_saving.cpp:
```diff
@@ -28,28 +28,51 @@ using namespace ls;
+ *    the "lambda injection" approach.
+ *    - Also, when a test breaks you need to spend a lot of time with the
+ *      debugger to figure out what went wrong.
+ *
```
These are reasonable suggestions, but I do think there's a place for high-level tests to supplement the unit tests, because it's too easy to leave something untested when creating unit tests alone, but higher-level end-to-end approaches test everything.  Ideally when something goes wrong, you find it in a unit test, but that doesn't always happen.  I would suggest that as we find bugs in our code that are only found by these higher-level end-to-end tests, that we add unit tests that would have found them.

CiaranWelsh replied:
>  there's a place for high-level tests to supplement the unit tests
Absolutely, but the unit tests must actually exist in order to be complemented by integration/end-to-end testing. I think I just wrote in frustration when I couldn't see what was broken let-a-lone what the problem actually was or how to fix it. 

> I would suggest that as we find bugs in our code that are only found by these higher-level end-to-end tests, that we add unit tests that would have found them

Seems like a good suggestion.

luciansmith commented on test/tmp.py:
```diff
@@ -0,0 +1,26 @@
+import sys
```
Again, this 'tmp' file looks suspicious.

CiaranWelsh replied:
Yep. Deleted.

luciansmith committed:
General cleanup from code review.

A lot of formatting is just 'whatever MSVS automatically did when I re-typed the final '}'.'
Also, remove Ciaran's personal error logging commands.

Something hinky is going on with RoadRunnerMapTests; the .cpp and .h files were virtually identical (and I made them exactly identical), which made it impossible to actually compile that test since everything was defined twice.  Commented out that test from CMakeLists for now.

luciansmith committed:
Don't write final SBML document.

Re-do of fix to not write SBML document after reading it in.

luciansmith committed:
Try to make tests link to dll not static library.

The tests themselves were taking up too much space on my hard drive, which is silly.  However, I'm not sure this actually fixes anything; they're still huge.

CiaranWelsh committed:
Remove rogue test

CiaranWelsh committed:
Delete header file -- not needed.

CiaranWelsh committed:
Delete tmp files

CiaranWelsh committed:
Add STS.h and STS.cpp.

CiaranWelsh committed:
Remove dead test

CiaranWelsh committed:
Clean up dead tests

CiaranWelsh committed:
Add performance test command line program written in C++ and a python script for running VTune with various runs.

CiaranWelsh committed:
Finshed first pass of RoadRunnerMap with docs and unit tests.

CiaranWelsh committed:
Merge branch 'orc-jit' of https://github.com/sys-bio/roadrunner into orc-jit

# Conflicts:
#	test/cxx_api_tests/CMakeLists.txt
#	test/cxx_api_tests/RoadRunnerMapTests.cpp
#	test/cxx_api_tests/RoadRunnerMapTests.h
#	test/llvm_tests/LLVMModelDataSymbolsTests.cpp

CiaranWelsh reviewed (commented):
This breaks the python tests on windows, since the TestModelFactory and other *pyd's now need to link the roadrunner.dll which is located under the bin directory. We could add the roadrunner.dll to the site-packages/roadrunner/tests directory but then we should probably link the main roadrunner module dynamically as well (or have two copies of the roadrunner code). I'm going to link the static version again for now so I can continue developing some python tests and we can make this a separate issue if you think necessary.

CiaranWelsh commented on test/CMakeLists.txt:
```diff
@@ -68,7 +68,7 @@ function(add_test_executable TEST_TARGET OUT_VARIABLE)
     # link to roadrunner-static, the flags needed should be transitive
     #    add_compile_definitions(${TEST_TARGET} PRIVATE STATIC_RR)
 
-    add_dependencies(${TEST_TARGET} roadrunner-static gtest gtest_main gmock gmock_main rr-mockups)
```
This breaks the python tests on windows, since the TestModelFactory and other *pyd's now need to link the roadrunner.dll which is located under the bin directory. We could add the roadrunner.dll to the site-packages/roadrunner/tests directory but then we should probably link the main roadrunner module dynamically as well (or have two copies of the roadrunner code). I'm going to link the static version again for now so I can continue developing some python tests and we can make this a separate issue if you think necessary.

luciansmith replied:
Ah, sorry.  Reverting it is fine.  The issue is just that with the increased size of the llvm binaries, these files are suddenly much larger than they were, and I was already at the edge of my hard drive size.  I'll move stuff around instead.

CiaranWelsh committed:
Fix incorrectly named python test: test_TestModelFactory.py to test_testModelFactory.py.

CiaranWelsh committed:
Added ability for imports to fail. This allows me to run python tests from the source tree, which has the significant advantage of your work not being overwritten by new build triggers (like if you develop in the build tree).

CiaranWelsh committed:
Added unique_ptr swig object. We don't want to actuall expose a wrapped std::unique_ptr<RoadRunner> but we do need it for some of the typemaps to work

CiaranWelsh committed:
Modifications to get RoadRunnerMap functional in python

CiaranWelsh committed:
RoadRunnerMap is wrapped by python and with unittests.

CiaranWelsh committed:
Fix test that broke becuase of the need to change return type of RoadRunnerMap::steal from unique_ptr to raw. This was needed because swig.

CiaranWelsh commented:
@luciansmith I'm more or less finished with `RoadRunnerMap` now. Would you mind reviewing the last few commits, paying particular attention to swig/python bits and memory management. Thanks.

CiaranWelsh committed:
Added calls to function in LLJit that results in creation of some llvm::Function's.

CiaranWelsh committed:
Added an assertion in the caching system for LLJit

CiaranWelsh committed:
Fixed caching test and disabled another since it is broken due to changes in sbml md5. This isn't an urgent fix however so leave disabled

CiaranWelsh committed:
Uncomment out jacobian and structural analysis tests. They seem to be working now?

CiaranWelsh committed:
Get rid of mutexes in read and write config file.

luciansmith commented on source/rrRoadRunnerOptions.cpp:
```diff
@@ -340,19 +334,94 @@ namespace rr {
+                break;
+            default:
+                std::string err = "Compiler option is invalid. Defaulting back to MCJit";
+                rrLogWarn << err;
```
I think I would make this an error, and might even throw an exception.  If someone said to do something and you don't, chances are they need to know.

CiaranWelsh replied:
Yep makes sense.

luciansmith commented on test/sbml-test-suite/STS.cpp:
```diff
@@ -0,0 +1,104 @@
+    return xmlFile.string();
+}
+
+std::vector<std::string> STS::getFirstNModelsFromSTS(int N, int start) {
```
I'm not sure how you're expecting this function to be used, but 'first N' seems highly specific.  Can you instead take just a list or a set or a range iterator as the argument, and return the tests that match the elements of that list?  Or just make it 'getModelNFromSTS' and return the single XML file, and let the user set up 'first N' or 'test x' or 'tests x-y' or 'tests x, y, and z' or 'all tests with the tag XXX', etc.?

CiaranWelsh replied:
I just needed something quickly to get going with [RoadRunnerMapTests.cpp](https://github.com/sys-bio/roadrunner/blob/80a3bc2c92648b131aa23f5091cc7a2b3b61739e/test/cxx_api_tests/RoadRunnerMapTests.cpp#L30) which loads models in parallel. Since I wasn't testing specific models or sbml features I didn't actually care about the identity of the models being loaded, hence "firstN". [We already have the [getModelNFromSTS](https://github.com/sys-bio/roadrunner/blob/56c78ae3b037eca6dfe210b1fc97faa2060b028b/test/sbml-test-suite/STS.h#L19), I'll refactor the name. 

The interface for this can definitely be improved. In general, I would like small but nice and easy to use couple of methods that provide access to the STS for unit/integration testing. The methods you've suggested make sense to me.

luciansmith replied:
Sure; how about we file an issue and leave it as-is for now?

CiaranWelsh replied:
I kind of accidently spent today on this. Actually I don't mind because its something I've wanted to do for a while. First draft will be checked in today -- feel free to modify as you see fit.

luciansmith replied:
Ha!  I understand that. Obviously not a problem; happy to look at the updated version.

luciansmith committed:
Remove the results.tsv file yet again, and ignore it.

luciansmith commented on wrappers/Python/roadrunner/roadrunner.i:
```diff
@@ -513,6 +511,51 @@
+    return newobj;
+}
+
+//%apply(std::unique_ptr<rr::RoadRunner>){
```
Residual code?

luciansmith commented on wrappers/Python/roadrunner/roadrunner.i:
```diff
@@ -513,6 +511,51 @@
+}
+
+
+//%typemap(in) RoadRunner*{
```
Residual code?  (If it's a helpful template, feel free to leave it in.)

CiaranWelsh committed:
throw instead of defaulting back to MCJit if user requests invalid llvm backend

CiaranWelsh committed:
delete stochastic results file

CiaranWelsh committed:
Merge branch 'orc-jit' of https://github.com/sys-bio/roadrunner into orc-jit

CiaranWelsh committed:
Spruced up STS class a bit.

Commit:
Build fixes on mac

Commit:
Undo use of name mangled symbols for functions loaded by MCJit or LLJit. This breaks their use where they are actually needed, since we use the unmangled versions.

Commit:
Do not use C++ API in C tests. This commit *may* have swept over a bug that occurs when you mix them.

luciansmith committed:
Don't need to include 'thread' twice.

luciansmith committed:
Merge branch 'orc-jit' of https://github.com/sys-bio/roadrunner into orc-jit

Commit:
Merge branch 'orc-jit' of https://github.com/sys-bio/roadrunner into orc-jit

Commit:
Minor commit, uncomment out some tests. Remove some rogue commited code.

Commit:
Okay, so this was a fun one. I'd noticed the other day that the tests link the roadrunner C api, whereas I thought they should be linking roadrunner-static (at least for now, in future as you said, it'd be better to use the shared version). When linking the roadrunner_c_api, moiety conservation works as expected on mac. However, when linking roadrunner-static, dynamic cast from SBasePlugin* to ConservedMoietyPlugin* fails. Changing this to a C style cast fixes the immediate nullptr problem but then fails with a "recursive symbol definition". Again, everything works fine on windows, this is a Mac problem (and maybe linux).

Commit:
Comment out RoadrunnerMapTests to see if everything else passes. They do on mac, locally.

Commit:
Added tests that catch a memory bug associated with rrSparse when rtying to RoadRunner::getIDs *after* a call to setConservedMoieties. I havent worked out how to fix this and so these tests are temporarily disabled. Issue #934 has been added to track this problem.

luciansmith commented:
The inexplicable part of this to me is how these problems arose in the first place.  At this point, would it be better to fork from the last working Azure build and re-incorporate the changes since then, but fixing each change to pass all the tests?

If the problems stem from making roadrunner itself parallel, the obvious solution is to revert that change, and make roadrunner parallel through the use of completely separate threads with save/loadState.

Commit:
Re-add all tests except for RoadRunnerMapTests. The latter works but needs moe effort with regards to thready safety and so will be developed further after orc-jit is merged into develop.

Commit:
disable RoadRunnerMap tests from Python.

Commit:
Added test in TestModelFactory that has events. This is a TimeSeriesResult type and I've added appropriate tests in CvodeIntegrationTest.cpp. This rules out the possibility that its the whoel event system that breaks in the stochastic test suite.

Commit:
Raise ValueError instead of self.fail, which is a syntax error.

Commit:
Fix bugs in distrib functions. Problem was with mixed up function signatures.

Commit:
CAPIModelEditingTests should be just ModelEditingTests

CiaranWelsh committed:
Revert CAPICoreTest.cpp to a working version. Not sure what broke but the revert puts it back to "normal".

CiaranWelsh committed:
try removing the -j4 option on manylinux, perhaps we have a deadlock which causes the manylinux build to hang?

CiaranWelsh committed:
updated llvm location on manylinux CI build, it has moved to a subdirectory of '/'

Commit:
Uncomment out RoadRunnerMapTests.cpp

Commit:
Add parallel_hashmap to source. This is an open source parallel hashmap implementation which we need for RoadRunnerMap to work. My own attempt at implementing a parallel hashmap didn't work so I now attempt to use this version. Notably, I've added the whole repository to the source directory. However, the implementation is still being actively developed so if all goes well we should add this as a dependency to libroadrunner-deps to make for easy updates.

CiaranWelsh committed:
try to exclude stochastic tests from the main test suite on manylinux and run it in its own Azure job.

CiaranWelsh committed:
try to exclude stochastic tests from the main test suite on manylinux and run it in its own Azure job. And fix typo

CiaranWelsh committed:
fix ctest command manylinux

CiaranWelsh committed:
fix ctest command manylinux

CiaranWelsh commented:
This branch actually passes all the tests now, verified in docker for many linux. The problem is with resource limitations on azure preventing the build from passing manylinux builds.

Commit:
Remove some files from parallel_hashmap that may interfere with project.

Commit:
Modifications to STS, RoadRunner and RoadRunnerMap until RoadRunnerMapTests.cpp pass with Asan and thread sanitizer. At first, I began basing this code off a std::unordered_map, which appeared to have some data races (according to thread sanitizer). I switched to parallel_hashmap, an open source project that I've added under source/parallel_hashmap and along with a few of my own locks, was able to pass the tests. Solver registration and llvm initialization were among the tasks that needed to be done under mutex protection.

Commit:
Add parallel hashmap as include directory in the build interface.

Commit:
Add orc-jit-roadrunner-container branch to azure

Commit:
update swig interface file with regards to changes in RoadRunnerMap. Python tests pass for RoadRunnerMap on Mac

Commit:
Modified dockerfile so that gcc gets built with -O3, perhaps this will speed the manylinux build?

CiaranWelsh committed:
add -O3 to dockerfile gcc build. Does it make a difference to manylinux build?

CiaranWelsh committed:
fix test that breaks on windows

CiaranWelsh committed:
fix llvm location on manylinux. Its been moved to /install-llvm-13.x

CiaranWelsh committed:
Merge branch 'orc-jit-rr-container' into orc-jit

CiaranWelsh committed:
disable running the stochastic test suite on manylinux. I've tried everything I can think of to speed this up (outside of modifying the code), including rebuilding gcc in the manylinux docker container with different optimization flags (-O3). One option is to run these tests in parallel on manylinux, but this requires further discussion. Disable running these tests for now so that I can move forward

CiaranWelsh committed:
Add download_biomodels.py to python subfolder. Copy it to roadrunner.tests, even though it is not a test.

CiaranWelsh committed:
added documentation for RoadRunnerMap

CiaranWelsh committed:
rebuild docs

CiaranWelsh committed:
Build and install valgrind in the manylinux dockerfile.

CiaranWelsh committed:
Make ctest verbose
